feat: ESG error handling continued

nsprenkle opened the pull request:
## Description

Add and extend error handling for ESG.

JIRA: https://openedx.atlassian.net/browse/AU-469
FYI: @edx/masters-devs-gta 

## Supporting information

⚠️ **I suggest reviewing with whitespace changes ignored!** There are a lot of indent changes due to blocks of code now living inside of try blocks, ignoring whitespace will make it more obvious of the _significant_ changes.

- Make internal XBlock handler data/exception handling more consistent: successful responses return data, failure responses raise exceptions, sometimes with context or special-case exception types for handling.
- Failures in an XBlock handler now return `HTTP 500: { 'error': 'ERR_INTERNAL' }` except established cases where we want to return extra data (lock/grade contest)
- Fall-through exceptions returned for uncaught and BFF-originating errors `HTTP 500: { 'error': 'ERR_UNKNOWN' }`
- Removes the `ERR_GRADE_SUBMIT` error code, returns `ERR_INTERNAL` for an ORA failure to submit grade.
- Moved ORA xblock handlers to a new file.
- Refactored some test logic.
- Updated GitHub Action test setup to include `ora_staff_grader` app in test counts.

## Testing instructions

Use example collection to exercise BFF endpoints, verify adherence to contracts. These include examples for failure states and returned exceptions.

nsprenkle committed:
chore: remove unused import

nsprenkle committed:
refactor: add exception with context

nsprenkle committed:
refactor: standardize base view handler exceptions

nsprenkle committed:
refactor: standardize view exception handling

All views largely operate inside a try/except to wrap and present errors
in a more consistent way to the frontend.

nsprenkle committed:
fix: update mock rubric data

nsprenkle committed:
test: refactor bad param testing

nsprenkle committed:
test: exception test updates

nsprenkle committed:
chore: update example collection

nsprenkle committed:
chore: pep8

nsprenkle committed:
chore: add ESG app to github actions test manifest

jansenk commented on .github/workflows/unit-test-shards.json:
```diff
@@ -4,7 +4,7 @@
   "lms-4": "lms/djangoapps/discussion/ lms/djangoapps/edxnotes/ lms/djangoapps/email_marketing/ lms/djangoapps/experiments/",
   "lms-5": "lms/djangoapps/gating/ lms/djangoapps/grades/ lms/djangoapps/instructor/ lms/djangoapps/instructor_analytics/",
-  "lms-6": "lms/djangoapps/instructor_task/ lms/djangoapps/learner_dashboard/ lms/djangoapps/lms_initialization/ lms/djangoapps/lms_xblock/ lms/djangoapps/lti_provider/ lms/djangoapps/mailing/ lms/djangoapps/mobile_api/ lms/djangoapps/monitoring/ lms/djangoapps/program_enrollments/ lms/djangoapps/rss_proxy/ lms/djangoapps/save_for_later/ lms/djangoapps/static_template_view/ lms/djangoapps/staticb ...
+  "lms-6": "lms/djangoapps/instructor_task/ lms/djangoapps/learner_dashboard/ lms/djangoapps/lms_initialization/ lms/djangoapps/lms_xblock/ lms/djangoapps/lti_provider/ lms/djangoapps/mailing/ lms/djangoapps/mobile_api/ lms/djangoapps/monitoring/ lms/djangoapps/ora_staff_grader/ lms/djangoapps/program_enrollments/ lms/djangoapps/rss_proxy/ lms/djangoapps/save_for_later/ lms/djangoapps/static_temp ...
```
whats this?

jansenk replied:
never mind! just read that slack convo

jansenk commented on lms/djangoapps/ora_staff_grader/tests/test_views.py:
```diff
@@ -216,42 +285,58 @@ def test_fetch_submission(
+    @patch('lms.djangoapps.ora_staff_grader.views.SubmissionFetchView.check_submission_lock')
+    def test_fetch_submission_exceptions(
+        self,
+        inject_chaos,
```
Nit: this param name sounds like a bool. `mock_exception_index` or something like that maybe?

jansenk commented on lms/djangoapps/ora_staff_grader/tests/test_views.py:
```diff
@@ -65,16 +65,94 @@ def url_with_params(self, params):
     view_name = 'ora-staff-grader:initialize'
 
-    def test_missing_ora_location(self):
+    # Options split for reuse
```
🧙‍♂️ 🧙‍♀️ 🦉 🦎 🪄 ⚡ 🧹

jansenk commented on lms/djangoapps/ora_staff_grader/views.py:
```diff
@@ -56,7 +64,17 @@ def get_rubric_config(self, request, usage_id):
+
         response_data = json.loads(response.content)
+
+        # Handled faillure still returns HTTP 200 but with 'success': False and supplied error message "msg"
```
nit: extra l in fail**l**ure

jansenk commented on lms/djangoapps/ora_staff_grader/views.py:
```diff
@@ -65,6 +83,10 @@ def get_submission_info(self, request, usage_id, submission_uuid):
         response = call_xblock_json_handler(request, usage_id, 'get_submission_info', data)
+
+        if response.status_code != 200:
+            raise Exception()
```
there are places in these handler where I raised JsonHandlerErrors with, say, 404 error codes. Will those get caught here? Is there some weird handler behavior that hides that? Some of them had error messages that I think would be useful to pass along.

nsprenkle replied:
This would catch any class of error and suppress error messaging. We _could_ add error information to the error context but I don't think it would actually be used by the frontend since it does its own error display based on BFF error classes. 

Might be helpful to add some sort of logging or analytics so we see what types of errors happen but the value of actually surfacing them to the frontend seems limited.

jansenk commented on lms/djangoapps/ora_staff_grader/views.py:
```diff
@@ -151,29 +191,35 @@ class InitializeView(StaffGraderBaseView):
+
+        # Catch bad ORA location
+        except (InvalidKeyError, ItemNotFoundError):
+            return BadOraLocationResponse()
```
Shouldn't this be on the other views that accept a location as well?

nsprenkle replied:
Not necessarily, there are 2 slightly different error paths:

1. This is the only endpoint that actually reads from the ORA item from the platform side. If it fails to find the ORA we need to know before we try to poll any of its data.
2. The other endpoints use XBlock handlers exclusively which, if passed a bad ORA location, return a 404. We could catch just those, but they would be through the XBlock handler wrapping code instead (to save us a modulestore lookup)

jansenk commented on lms/djangoapps/ora_staff_grader/views.py:
```diff
@@ -82,14 +108,30 @@ def submit_grade(self, request, usage_id, grade_data):
+
+        # Unclear that there would every be an error (except network/auth) but good to catch here
+        if response.status_code != 200:
+            raise Exception()
```
I think this should be a custom error we're raising in all these places so we can avoid the flat `except Exception:` in the views. We want to know when this scenario happens, but other exceptions that may be raised should probably still bubble up so we can catch them in tools like NewRelic.

jansenk reviewed (changes requested):
Some Qs, some changes, mostly great though

nsprenkle committed:
refactor: move ORA handlers into separate file

nsprenkle committed:
refactor: XBlock handler exception refactor

XBlock handlers return XBlockInternalError with handler name and
optional additional context for blanket ORA errors

nsprenkle committed:
refactor: remove submit grade error, add internal

nsprenkle committed:
fix: add missing handler info to failure in submit

nsprenkle committed:
refactor: update error handling in ESG views

All views have the possiblity of returning an XBlockInternalError for
issues encountered in an XBlock handler as well as an UnknownError for
exceptions in the view itself e.g. broken data, fall-through exceptions

nsprenkle committed:
test: update view tests

nsprenkle committed:
refactor: move common test data to a separate file

nsprenkle committed:
refactor: reduce log in boilerplate

nsprenkle committed:
chore: pep8
